Changing linkage type mapping

AlexeySotkin opened the pull request:
Linkage type mapping between LLVM and SPIRV is as follows:
llvm private/internal linkage -> SPIRV Internal -> llvm internal linkage
llvm global value definitions with available_externally linkage or with common linkage -> SPIRV Import -> llvm available_externally linkage
llvm global value declaration(w/o available_externally or common linkage ) -> SPIRV Import -> llvm external linkage
other llvm global values  -> SPIRV Export -> llvm external linkage

Print readable names for linkage names and linkage types in text mode 

Bumping up translator version

AlexeySotkin committed:
Changing linkage type mapping.
Since global variables and functions with external linkage in C code
can be mapped to either import or export linkage type in SPIR-V,
new approach has been implemented:
- global variables and function declarations are to be decorated with Import
 linkage type in SPIR-V.
- global variables and function definitions are to be decorated with Export
 linkage type in SPIR-V.
- variables and functions with internal linkage type will have default linkage
 type in SPIR-V, i.e. internal.

Conflicts:
	lib/SPIRV/SPIRVWriter.cpp

AlexeySotkin committed:
Linkage type mapping between LLVM and SPIRV is as follows:
llvm private/internal linkage -> SPIRV Internal -> llvm internal linkage
llvm global value definitions with available_externally linkage or with common linkage -> SPIRV Import -> llvm available_externally linkage
llvm global value declaration(w/o available_externally or common linkage ) -> SPIRV Import -> llvm external linkage
other llvm global values  -> SPIRV Export -> llvm external linkage

AlexeySotkin committed:
Print readable names for linkage names and linkage types in text mode
Update LIT tests for linkage types

AlexeySotkin committed:
Merge branch 'github/spirv-3.6.1' into changing-linkage-type-mapping

AlexeySotkin committed:
Bumping up translator version

AlexeySotkin committed:
Merge branch 'github/spirv-3.6.1' into changing-linkage-type-mapping

pierremoreau commented:
So this should fix #6 I guess: I'll give it a try later today.

yxsamliu commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -725,7 +725,7 @@ LLVMToSPIRV::transValueWithoutDecoration(Value *V, SPIRVBasicBlock *BB,
     auto BVar = static_cast<SPIRVVariable *>(BM->addVariable(
         transType(GV->getType()), GV->isConstant(),
-        SPIRSPIRVLinkageTypeMap::map(GV->getLinkage()),
+        transLinkageType(static_cast<GlobalValue*>(GV)),
```
GV is of type GlobalVariable, do we still need this cast?

AlexeySotkin replied:
No. Fixed.

yxsamliu commented on lib/SPIRV/libSPIRV/SPIRVDecorate.h:
```diff
@@ -149,6 +150,11 @@ class SPIRVDecorateLinkageAttr:public SPIRVDecorate{
   }
+
+  void encode(std::ostream &O)const {
+    getEncoder(O) << Target << Dec << getLinkageName() << SPIRVTextOnly(" ")
```
should only do this if SPIRVUseTextFormat is true, otherwise still need output binary format.

Also, need corresponding decode() function.

AlexeySotkin replied:
Fixed

pierremoreau commented:
It does fix #6, great!

AlexeySotkin committed:
Applying comments after code review:
Unncessary casts are removed.
Fixed encoding/decoding of linkage attribute literals.

AlexeySotkin committed:
Guarding textual output with ifdef macro

yxsamliu commented:
sorry for the delay. Since github did not notify new commits in pull request, I just saw your last commit. It looks good to me. However it seems there are conflicts need to be resolved before it can be merged, probably due to last merge.

AlexeySotkin committed:
Merge remote-tracking branch 'origin/github/spirv-3.6.1' into changing-linkage-type-mapping

Conflicts:
	lib/SPIRV/libSPIRV/SPIRVDecorate.cpp
